File: 07-cli-browser/http-client.go
package main

import (
	"fmt"
	"io"
	"net"
	"os"
)

func main() {
	if len(os.Args) < 2 {
		fmt.Println("Usage: go run http-client.go <host>")
		fmt.Println("Example: go run http-client.go example.com")
		return
	}

	host := os.Args[1]

	// Step 1: TCP connection
	fmt.Printf("Connecting to %s:80...\n", host)
	conn, err := net.Dial("tcp", host+":80")
	if err != nil {
		fmt.Println("Connection error:", err)
		return
	}
	defer conn.Close()
	fmt.Println("Connected!")

	// Step 2: Build HTTP request
	request := fmt.Sprintf(
		"GET / HTTP/1.1\r\nHost: %s\r\nConnection: close\r\n\r\n",
		host,
	)

	// Step 3: Send request
	fmt.Printf("\nSending request:\n%s\n", request)
	_, err = conn.Write([]byte(request))
	if err != nil {
		fmt.Println("Write error:", err)
		return
	}

	// Step 4: Read response
	fmt.Println("Response:")
	fmt.Println("--------")
	body, err := io.ReadAll(conn)
	if err != nil {
		fmt.Println("Read error:", err)
		return
	}

	fmt.Println(string(body))
}

File: 07-cli-browser/http-client.py
import socket
import sys

def main():
    if len(sys.argv) < 2:
        print("Usage: python http-client.py <host>")
        print("Example: python http-client.py example.com")
        return

    host = sys.argv[1]

    # Step 1: TCP connection
    print(f"Connecting to {host}:80...")
    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    sock.connect((host, 80))
    print("Connected!")

    # Step 2: Build HTTP request
    request = (
        f"GET / HTTP/1.1\r\n"
        f"Host: {host}\r\n"
        f"Connection: close\r\n"
        f"\r\n"
    )

    # Step 3: Send request
    print(f"\nSending request:\n{request}")
    sock.send(request.encode())

    # Step 4: Read full response
    print("Response:")
    print("--------")

    chunks = []
    while True:
        data = sock.recv(4096)
        if not data:
            break
        chunks.append(data)

    response = b"".join(chunks)
    print(response.decode(errors="ignore"))

    sock.close()

if __name__ == "__main__":
    main()
